pandas/_libs/index.pyx

```diff
@@ -599,70 +600,72 @@ cpdef convert_scalar(ndarray arr, object value):
     return value
 
 
-cdef class MultiIndexObjectEngine(ObjectEngine):
-    """
-    provide the same interface as the MultiIndexEngine
-    but use the IndexEngine for computation
+cdef class BaseMultiIndexCodesEngine(object):
+    def __init__(self, levels, labels, offsets, **kwargs):
+        self._levels = levels
```

Comment:
these don't need to be private (levels & offsets)

you *can* type them (offsets is an uint64_t array?)

furthermore these can be made readonly as well (you can set them once), see how we do this in Timestamp for example

Reply:
If I add
```python
    cdef readonly:
        object levels
        ndarray offsets
```
I get
```python
In [1]: import pandas as pd
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-7dd3504c366f> in <module>()
----> 1 import pandas as pd

/home/nobackup/repo/pandas/pandas/__init__.py in <module>()
     40 import pandas.core.config_init
     41 
---> 42 from pandas.core.api import *
     43 from pandas.core.sparse.api import *
     44 from pandas.stats.api import *

/home/nobackup/repo/pandas/pandas/core/api.py in <module>()
      8 from pandas.core.dtypes.missing import isna, isnull, notna, notnull
      9 from pandas.core.categorical import Categorical
---> 10 from pandas.core.groupby import Grouper
     11 from pandas.io.formats.format import set_eng_float_format
     12 from pandas.core.index import (Index, CategoricalIndex, Int64Index,

/home/nobackup/repo/pandas/pandas/core/groupby.py in <module>()
     46 from pandas.core.base import (PandasObject, SelectionMixin, GroupByError,
     47                               DataError, SpecificationError)
---> 48 from pandas.core.index import (Index, MultiIndex,
     49                                CategoricalIndex, _ensure_index)
     50 from pandas.core.categorical import Categorical

/home/nobackup/repo/pandas/pandas/core/index.py in <module>()
      1 # flake8: noqa
----> 2 from pandas.core.indexes.api import *
      3 from pandas.core.indexes.multi import _sparsify

/home/nobackup/repo/pandas/pandas/core/indexes/api.py in <module>()
      6                                       InvalidIndexError)  # noqa
      7 from pandas.core.indexes.category import CategoricalIndex  # noqa
----> 8 from pandas.core.indexes.multi import MultiIndex  # noqa
      9 from pandas.core.indexes.interval import IntervalIndex  # noqa
     10 from pandas.core.indexes.numeric import (NumericIndex, Float64Index,  # noqa

/home/nobackup/repo/pandas/pandas/core/indexes/multi.py in <module>()
     52 
     53 class MultiIndexUIntEngine(libindex.BaseMultiIndexCodesEngine,
---> 54                            libindex.UInt64Engine):
     55     """
     56     Manage a MultiIndex by mapping label combinations to positive integers.

TypeError: multiple bases have instance lay-out conflict
```